commit 5f4e1908d910
Author: Boris Zbarsky <bzbarsky@mit.edu>
Date:   Thu Dec 14 12:45:37 2017 -0500

    Bug 1425107 - Fix setting scrollTop to 0 on an element with a pending scroll position restore to actually work correctly. r=mats, a=gchang
    
    MozReview-Commit-ID: 949eBXmKHlA
    
    --HG--
    extra : source : 81174f568eaa91aa8f5f65e05f35f4350a8cfb25
    extra : amend_source : a1142868a5029bcbaea6fc66bc0becd3ef8d6c9b
---
 layout/generic/nsGfxScrollFrame.cpp                |  7 ++++++
 testing/web-platform/meta/MANIFEST.json            | 27 +++++++++++++++++++++-
 .../cssom-view/scrollTop-display-change-ref.html   |  8 +++++++
 .../css/cssom-view/scrollTop-display-change.html   | 17 ++++++++++++++
 4 files changed, 58 insertions(+), 1 deletion(-)

diff --git layout/generic/nsGfxScrollFrame.cpp layout/generic/nsGfxScrollFrame.cpp
index 9fee3cc2e502..b1575b214240 100644
--- layout/generic/nsGfxScrollFrame.cpp
+++ layout/generic/nsGfxScrollFrame.cpp
@@ -2272,6 +2272,11 @@ ScrollFrameHelper::ScrollToWithOrigin(nsPoint aScrollPosition,
                                           const nsRect* aRange,
                                           nsIScrollbarMediator::ScrollSnapMode aSnap)
 {
+  if (aOrigin != nsGkAtoms::restore) {
+    // If we're doing a non-restore scroll, we don't want to later
+    // override it by restoring our saved scroll position.
+    mRestorePos.x = mRestorePos.y = -1;
+  }
 
   if (aSnap == nsIScrollableFrame::ENABLE_SNAP) {
     GetSnapPointForDestination(nsIScrollableFrame::DEVICE_PIXELS,
@@ -4316,6 +4321,8 @@ ScrollFrameHelper::ScrollToRestoredPosition()
           (mScrollPort.XMost() - scrollToPos.x - mScrolledFrame->GetRect().width);
       }
       AutoWeakFrame weakFrame(mOuter);
+      // It's very important to pass nsGkAtoms::restore here, so
+      // ScrollToWithOrigin won't clear out mRestorePos.
       ScrollToWithOrigin(scrollToPos, nsIScrollableFrame::INSTANT,
                          nsGkAtoms::restore, nullptr);
       if (!weakFrame.IsAlive()) {
